### PR TITLE
🥔✨ `Journal`: extract `Term`s from `Entry`s

### PR DESCRIPTION
**Pull request by zspencer**
- https://github.com/zinc-collective/convene/issues/1566

Not yet linking to them; but that should be a decent next step.

### CODE REVIEW

File: app/furniture/journal/entry.rb [hunk 1]
```diff
@@ -24,8 +24,11 @@ class Entry < ApplicationRecord
 
     # @!attribute journal
     #   @return [Journal::Journal]
-    belongs_to :journal, class_name: "Journal::Journal", inverse_of: :entries
-    delegate :room, :space, to: :journal
+    belongs_to :journal, inverse_of: :entries
+    has_one :room, through: :journal
+    has_one :space, through: :journal
```
Comment (zspencer) [inline app/furniture/journal/entry.rb:29]: This was some minor cleanup from some poorly-connected linkages in the object graph. Happy to split this out.

File: app/furniture/journal/term.rb [hunk 2]
```diff
@@ -0,0 +1,9 @@
+class Journal
+  class Term < ApplicationRecord
+    location(parent: :journal)
+
+    self.table_name = "journal_terms"
+    validates :canonical_term, presence: true
```
Comment (zspencer) [inline app/furniture/journal/term.rb:6]: Probably should also validate uniqueness scoped to the journal.

File: app/furniture/journal/entry.rb [hunk 3]
```diff
@@ -46,6 +49,15 @@ def self.renderer
       )
     end
 
+    def extract_terms
+      body.scan(/#(\w+)/)&.flatten&.each do |term|
+        next if journal.terms.where("aliases ILIKE ?", term)
+          .or(journal.terms.where(canonical_term: term)).exists?
+
+        journal.terms.find_or_create_by!(canonical_term: term)
+      end
```
Comment (zspencer) [inline app/furniture/journal/entry.rb:58]: TBH this method feels like it wants to be a factory method on `Term`, something like `Term.extract(body, journal: journal)`  (unless we can get `journal.terms.extract(body)` to work; which would be even better)

File: app/furniture/journal/renderer.rb [hunk 4]
```diff
@@ -6,7 +6,9 @@ def autolink(link, link_type)
     end
 
     def postprocess(doc)
-      doc.gsub(/@([a-zA-Z\d]*)@(.*\.[a-zA-Z]*)/, '<a href="https://\2/@\1">@\1@\2</a>')
+      doc
+        .gsub(/@([a-zA-Z\d]*)@(.*\.[a-zA-Z]*)/, '<a href="https://\2/@\1">@\1@\2</a>')
+      # .gsub(/#(\w+)/, '<a href="../terms/\1">#\1</a>')
```
Comment (zspencer) [inline app/furniture/journal/renderer.rb:11]: Leftovers; just wanted to commit it before I fell asleep.

**Comment by anaulin** [comment]
Are these meant to be... tags? keywords? something else? I feel like using more "standard" words from the blogging world might make it easier for both coders and people using the app.

File: app/furniture/journal/renderer.rb (hunk 4 above)
Comment (anaulin) [reply app/furniture/journal/renderer.rb:11]: Should this commented-out code be removed, or are you keeping it for the future?

File: db/schema.rb [hunk 5]
```diff
@@ -117,6 +117,15 @@
     t.index ["journal_id"], name: "index_journal_entries_on_journal_id"
   end
 
+  create_table "journal_terms", id: :uuid, default: -> { "gen_random_uuid()" }, force: :cascade do |t|
+    t.uuid "journal_id"
+    t.string "canonical_term"
+    t.text "aliases"
```
Comment (anaulin) [inline db/schema.rb:123]: Would be nice to make this an array, or something more structured than just text (which I assume right now it means to be comma-separated?).

**Comment by anaulin** [comment]
Oh, I see now that you are extracting hashtags from the body. So maybe "tags" would be a better name?

**Review by anaulin — APPROVED** [review]
Nothing blocking, but I would love for this new thing to be named in a more "standard" way, e.g. "tags" or "keyword".

File: db/schema.rb (hunk 5 above)
Comment (zspencer) [reply db/schema.rb:123]: I'll give it a shot! I have not used array types at the database level before but it shouldn't be too hard.

**Comment by zspencer** [comment]
I like `Keyword`! I just didn't want to use `Tag` or `Hashtag`

File: app/furniture/journal/renderer.rb (hunk 4 above)
Comment (zspencer) [reply app/furniture/journal/renderer.rb:11]: It should not be in this branch; but the follow on one...

**Pull request merged**